### PR TITLE
Add one test for plugin type to `PluginsLoaderTests`

### PR DESCRIPTION
**Pull request by ldematte**
Add one test per plugin type (modular, stable, non-modular) to `PluginsLoaderTests`.

We split responsibility of `PluginsService` in a separate `PluginsLoader`; the `PluginsLoader` now is responsible to find and create all plugins bundles and descriptors (which is delegated to `PluginsUtils`, and partially covered by `PluginsUtilsTests`), and particularly to create all the ClassLoaders and Layers that will be used to load the plugin classes and create the plugin objects.

Currently, `PluginsLoader` functions are tested (in part) by `PluginsServiceTests`; this PR borrows from these tests (with partial duplication, ATM) and adds new ones to `PluginsLoaderTests`; the new tests focus on the responsibilities of `PluginsLoader` (find the jars, build the descriptors, build the layers), covering all 3 plugin types we can create.

Tests in `PluginsServiceTests` should probably be re-worked to remove duplication, and focus more on what responsibilities remain in `PluginsService` (actual plugin loading/creation, filtering, holding the registry), but I'd like to work incrementally and do that in a separate PR.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/plugins/PluginsLoader.java [hunk 1]
```diff
@@ -64,18 +64,26 @@ public interface PluginLayer {
...
-     * @param spiClassLoader The exported {@link ClassLoader} visible to other Java modules
-     * @param spiModuleLayer The exported {@link ModuleLayer} visible to other Java modules
+     * @param pluginModuleLayer The {@link ModuleLayer} containing the Java modules of the plugin
+     * @param spiClassLoader    The exported {@link ClassLoader} visible to other Java modules
+     * @param spiModuleLayer    The exported {@link ModuleLayer} visible to other Java modules
      */
     private record LoadedPluginLayer(
         PluginBundle pluginBundle,
         ClassLoader pluginClassLoader,
+        ModuleLayer pluginModuleLayer,
```
Comment (ldematte) [inline server/src/main/java/org/elasticsearch/plugins/PluginsLoader.java:86]: Fear not, this is not added here just for this test: we need it anyway/it's one of the reasons we wanted to split PluginsLoader out (see it in use in https://github.com/elastic/elasticsearch/pull/117239)

File: server/src/main/java/org/elasticsearch/plugins/PluginsLoader.java [hunk 2]
```diff
@@ -389,7 +406,7 @@ static String toModuleName(String name) {
         return result;
     }
 
-    static final String toPackageName(String className) {
+    static String toPackageName(String className) {
```
Comment (ldematte) [inline server/src/main/java/org/elasticsearch/plugins/PluginsLoader.java:409]: Automatic cleanup (IJ complaining)

Comment (jdconrad) [reply server/src/main/java/org/elasticsearch/plugins/PluginsLoader.java:419]: I'm surprised that wasn't caught before.

File: server/src/test/java/org/elasticsearch/plugins/PluginsLoaderTests.java [hunk 3]
```diff
@@ -28,4 +61,220 @@ public void testToModuleName() {
...
+            "version",
+            "1.0.0",
+            "elasticsearch.version",
+            Version.CURRENT.toString(),
+            "java.version",
+            System.getProperty("java.specification.version")
+        );
+
+        Path jar = plugin.resolve("impl.jar");
+        JarUtils.createJarWithEntries(jar, Map.of("p/A.class", InMemoryJavaCompiler.compile("p.A", """
```
Comment (jdconrad) [inline server/src/test/java/org/elasticsearch/plugins/PluginsLoaderTests.java:86]: This is quite clever to use in testing :)

**Review by jdconrad — APPROVED** [review]
LGTM! Thanks for adding these additional tests.

**Pull request merged**